File: cpp/src/IceDiscovery/LocatorI.cpp
// Copyright (c) ZeroC, Inc.

#include "LocatorI.h"
#include "../Ice/Random.h"
#include "Ice/Communicator.h"
#include "Ice/LocalExceptions.h"
#include "Ice/ObjectAdapter.h"
#include "LookupI.h"

#include <iterator>

using namespace std;
using namespace Ice;
using namespace IceDiscovery;

LocatorRegistryI::LocatorRegistryI(const Ice::CommunicatorPtr& communicator)
    : _wellKnownProxy{
          ObjectPrx{communicator, "dummy"}.ice_locator(nullopt).ice_router(nullopt).ice_collocationOptimized(true)}
{
}

void
LocatorRegistryI::setAdapterDirectProxy(string adapterId, optional<ObjectPrx> proxy, const Ice::Current&)
{
    lock_guard lock(_mutex);
    if (proxy)
    {
        _adapters.insert({adapterId, std::move(*proxy)});
    }
    else
    {
        _adapters.erase(adapterId);
    }
}

void
LocatorRegistryI::setReplicatedAdapterDirectProxy(
    string adapterId,
    string replicaGroupId,
    optional<ObjectPrx> proxy,
    const Ice::Current&)
{
    lock_guard lock(_mutex);
    if (proxy)
    {
        _adapters.insert({adapterId, std::move(*proxy)});
        auto p = _replicaGroups.find(replicaGroupId);
        if (p == _replicaGroups.end())
        {
            p = _replicaGroups.insert(make_pair(replicaGroupId, set<string>())).first;
        }
        p->second.insert(adapterId);
    }
    else
    {
        _adapters.erase(adapterId);
        auto p = _replicaGroups.find(replicaGroupId);
        if (p != _replicaGroups.end())
        {
            p->second.erase(adapterId);
            if (p->second.empty())
            {
                _replicaGroups.erase(p);
            }
        }
    }
}

void
LocatorRegistryI::setServerProcessProxy(string, optional<ProcessPrx>, const Ice::Current&)
{
}

optional<Ice::ObjectPrx>
LocatorRegistryI::findObject(const Ice::Identity& id) const
{
    lock_guard lock(_mutex);
    if (id.name.empty())
    {
        return nullopt;
    }

    Ice::ObjectPrx prx = _wellKnownProxy->ice_identity(id);

    vector<string> adapterIds;
    for (const auto& [adapterId, _] : _replicaGroups)
    {
        try
        {
            prx->ice_adapterId(adapterId)->ice_ping();
            adapterIds.push_back(adapterId);
        }
        catch (const Ice::Exception&)
        {
            // Ignore
        }
    }

    if (adapterIds.empty())
    {
        for (const auto& [adapterId, _] : _adapters)
        {
            try
            {
                prx->ice_adapterId(adapterId)->ice_ping();
                adapterIds.push_back(adapterId);
            }
            catch (const Ice::Exception&)
            {
                // Ignore
            }
        }
    }

    if (adapterIds.empty())
    {
        return nullopt;
    }

    IceInternal::shuffle(adapterIds.begin(), adapterIds.end());
    return prx->ice_adapterId(adapterIds[0]);
}

optional<Ice::ObjectPrx>
LocatorRegistryI::findAdapter(const string& adapterId, bool& isReplicaGroup) const
{
    lock_guard lock(_mutex);

    auto p = _adapters.find(adapterId);
    if (p != _adapters.end())
    {
        isReplicaGroup = false;
        return p->second;
    }

    auto q = _replicaGroups.find(adapterId);
    if (q != _replicaGroups.end())
    {
        Ice::EndpointSeq endpoints;
        optional<Ice::ObjectPrx> prx;
        for (const auto& r : q->second)
        {
            auto s = _adapters.find(r);
            if (s == _adapters.end())
            {
                continue; // TODO: Inconsistency
            }

            if (!prx)
            {
                prx = s->second;
            }

            Ice::EndpointSeq endpts = s->second->ice_getEndpoints();
            copy(endpts.begin(), endpts.end(), back_inserter(endpoints));
        }

        if (prx)
        {
            isReplicaGroup = true;
            return prx->ice_endpoints(endpoints);
        }
    }

    isReplicaGroup = false;
    return nullopt;
}

LocatorI::LocatorI(LookupIPtr lookup, LocatorRegistryPrx registry)
    : _lookup(std::move(lookup)),
      _registry(std::move(registry))
{
}

void
LocatorI::findObjectByIdAsync(
    Ice::Identity id,
    function<void(const optional<ObjectPrx>&)> response,
    function<void(exception_ptr)> ex,
    const Ice::Current&) const
{
    _lookup->findObject(make_pair(response, ex), id);
}

void
LocatorI::findAdapterByIdAsync(
    string adapterId,
    function<void(const optional<ObjectPrx>&)> response,
    function<void(exception_ptr)> ex,
    const Ice::Current&) const
{
    _lookup->findAdapter(make_pair(response, ex), adapterId);
}

void
LocatorI::getRegistryAsync(
    function<void(const optional<LocatorRegistryPrx>&)> response,
    function<void(exception_ptr)>,
    const Ice::Current&) const
{
    response(_registry);
}

File: csharp/src/IceDiscovery/LocatorI.cs
// Copyright (c) ZeroC, Inc.

namespace IceDiscovery;

internal class LocatorRegistryI : Ice.LocatorRegistryDisp_
{
    public
    LocatorRegistryI(Ice.Communicator com)
    {
        _wellKnownProxy = com.stringToProxy("p").ice_locator(null).ice_router(null).ice_collocationOptimized(true);
    }

    public override Task
    setAdapterDirectProxyAsync(string adapterId, Ice.ObjectPrx proxy, Ice.Current current)
    {
        lock (_mutex)
        {
            if (proxy != null)
            {
                _adapters[adapterId] = proxy;
            }
            else
            {
                _adapters.Remove(adapterId);
            }
        }
        return Task.CompletedTask;
    }

    public override Task
    setReplicatedAdapterDirectProxyAsync(
        string adapterId,
        string replicaGroupId,
        Ice.ObjectPrx proxy,
        Ice.Current current)
    {
        lock (_mutex)
        {
            if (proxy != null)
            {
                _adapters[adapterId] = proxy;
                HashSet<string> adapterIds;
                if (!_replicaGroups.TryGetValue(replicaGroupId, out adapterIds))
                {
                    adapterIds = new HashSet<string>();
                    _replicaGroups.Add(replicaGroupId, adapterIds);
                }
                adapterIds.Add(adapterId);
            }
            else
            {
                _adapters.Remove(adapterId);
                HashSet<string> adapterIds;
                if (_replicaGroups.TryGetValue(replicaGroupId, out adapterIds))
                {
                    adapterIds.Remove(adapterId);
                    if (adapterIds.Count == 0)
                    {
                        _replicaGroups.Remove(replicaGroupId);
                    }
                }
            }
        }
        return Task.CompletedTask;
    }

    public override Task
    setServerProcessProxyAsync(string id, Ice.ProcessPrx process, Ice.Current current)
    {
        return Task.CompletedTask;
    }

    internal Ice.ObjectPrx findObject(Ice.Identity id)
    {
        lock (_mutex)
        {
            if (id.name.Length == 0)
            {
                return null;
            }

            Ice.ObjectPrx prx = _wellKnownProxy.ice_identity(id);

            List<string> adapterIds = new List<string>();
            foreach (KeyValuePair<string, HashSet<string>> entry in _replicaGroups)
            {
                try
                {
                    prx.ice_adapterId(entry.Key).ice_ping();
                    adapterIds.Add(entry.Key);
                }
                catch (Ice.Exception)
                {
                }
            }
            if (adapterIds.Count == 0)
            {
                foreach (KeyValuePair<string, Ice.ObjectPrx> entry in _adapters)
                {
                    try
                    {
                        prx.ice_adapterId(entry.Key).ice_ping();
                        adapterIds.Add(entry.Key);
                    }
                    catch (Ice.Exception)
                    {
                    }
                }
            }

            if (adapterIds.Count == 0)
            {
                return null;
            }
            Ice.UtilInternal.Collections.Shuffle(ref adapterIds);
            return prx.ice_adapterId(adapterIds[0]);
        }
    }

    internal Ice.ObjectPrx findAdapter(string adapterId, out bool isReplicaGroup)
    {
        lock (_mutex)
        {
            Ice.ObjectPrx result = null;
            if (_adapters.TryGetValue(adapterId, out result))
            {
                isReplicaGroup = false;
                return result;
            }

            HashSet<string> adapterIds;
            if (_replicaGroups.TryGetValue(adapterId, out adapterIds))
            {
                List<Ice.Endpoint> endpoints = new List<Ice.Endpoint>();
                foreach (string a in adapterIds)
                {
                    Ice.ObjectPrx proxy;
                    if (!_adapters.TryGetValue(a, out proxy))
                    {
                        continue; // TODO: Inconsistency
                    }

                    result ??= proxy;

                    endpoints.AddRange(proxy.ice_getEndpoints());
                }

                if (result != null)
                {
                    isReplicaGroup = true;
                    return result.ice_endpoints(endpoints.ToArray());
                }
            }

            isReplicaGroup = false;
            return null;
        }
    }

    private readonly Ice.ObjectPrx _wellKnownProxy;
    private readonly Dictionary<string, Ice.ObjectPrx> _adapters = new Dictionary<string, Ice.ObjectPrx>();
    private readonly Dictionary<string, HashSet<string>> _replicaGroups = new Dictionary<string, HashSet<string>>();
    private readonly object _mutex = new();
}

internal class LocatorI : Ice.LocatorDisp_
{
    public LocatorI(LookupI lookup, Ice.LocatorRegistryPrx registry)
    {
        _lookup = lookup;
        _registry = registry;
    }

    public override Task<Ice.ObjectPrx>
    findObjectByIdAsync(Ice.Identity id, Ice.Current current)
    {
        return _lookup.findObject(id);
    }

    public override Task<Ice.ObjectPrx>
    findAdapterByIdAsync(string adapterId, Ice.Current current)
    {
        return _lookup.findAdapter(adapterId);
    }

    public override Ice.LocatorRegistryPrx getRegistry(Ice.Current current)
    {
        return _registry;
    }

    private readonly LookupI _lookup;
    private Ice.LocatorRegistryPrx _registry;
}
